lib/ExtUtils/t/Embed.t: failures on OpenBSD when compiling with gcc but not with clang

For several months I have been noticing smoke-test failures under certain configurations on OpenBSD-7.4; see, *e.g.,* [report 5052514](https://perl5.test-smoke.org/report/5052514).  The failures occur when configuring a perl built with these command-line switches:
```
$ sh ./Configure -des -Dusedevel -Dcc=gcc -Accflags='-DPERL_RC_STACK -DDEBUG_LEAKING_SCALARS'
```
... and then, after running `make test_prep`, calling:
```
$ cd t; ./perl harness -v ../lib/ExtUtils/t/Embed.t; cd - 
ld: error: undefined symbol: Perl_more_sv
>>> referenced by embed_test.c
>>>               /tmp//ccE9t92Z.o:(S_new_SV)

ld: error: undefined symbol: PL_sv_serial
>>> referenced by embed_test.c
>>>               /tmp//ccE9t92Z.o:(S_new_SV)
>>> referenced by embed_test.c
>>>               /tmp//ccE9t92Z.o:(S_new_SV)
collect2: ld returned 1 exit status

not ok 1
not ok 10 # system returned -1
Failed 10/10 subtests 

Test Summary Report
-------------------
../lib/ExtUtils/t/Embed.t (Wstat: 0 Tests: 2 Failed: 2)
  Failed tests:  1, 10
  Parse errors: Tests out of sequence.  Found (10) but expected (2)
                Bad plan.  You planned 10 tests but ran 2.
Files=1, Tests=2,  0 wallclock secs ( 0.02 usr  0.01 sys +  0.48 cusr  0.11 csys =  0.62 CPU)
Result: FAIL
```
I didn't pay these much attention at first because [at first](https://perl5.test-smoke.org/report/5049575) the perl was being built with a compiler I had never heard of, `eg++`.  But more recently these failures were occuring with `gcc` as the compiler.

I haven't been able to install an OpenBSD more recent than 6.9, but I decided to explore this problem on that version today regardless.  I got the test failure reported above on blead.  I then decided to see whether these failures were due to a recent code change or whether @cjg-cguevara's exploration of these config options had merely revealed a problem that was "always" there.  I tested at tag `v5.38.0` and got the failure reported above.
```
$ uname -mrs
OpenBSD 6.9 amd64

$ gcc --version | head -1
gcc (GCC) 4.2.1 20070719

$ git show | head -3
commit 76298ae68aa7796f0ffc05095b127d23f4b2de8f
Author: Ricardo Signes <rjbs@semiotic.systems>
Date:   Sat Jul 1 20:48:27 2023 -0400

$ sh ./Configure -des -Dusedevel -Dcc=gcc -Accflags='-DPERL_RC_STACK -DDEBUG_LEAKING_SCALARS'; make test_prep

$ ./perl -Ilib -V:config_args
config_args='-des -Dusedevel -Dcc=gcc -Accflags=-DPERL_RC_STACK -DDEBUG_LEAKING_SCALARS';

$ cd t; ./perl harness -v ../lib/ExtUtils/t/Embed.t; cd - 
ld: error: undefined symbol: Perl_more_sv
>>> referenced by embed_test.c
>>>               /tmp//ccE9t92Z.o:(S_new_SV)

ld: error: undefined symbol: PL_sv_serial
>>> referenced by embed_test.c
>>>               /tmp//ccE9t92Z.o:(S_new_SV)
>>> referenced by embed_test.c
>>>               /tmp//ccE9t92Z.o:(S_new_SV)
collect2: ld returned 1 exit status

not ok 1
not ok 10 # system returned -1
Failed 10/10 subtests 

Test Summary Report
-------------------
../lib/ExtUtils/t/Embed.t (Wstat: 0 Tests: 2 Failed: 2)
  Failed tests:  1, 10
  Parse errors: Tests out of sequence.  Found (10) but expected (2)
                Bad plan.  You planned 10 tests but ran 2.
Files=1, Tests=2,  0 wallclock secs ( 0.02 usr  0.01 sys +  0.48 cusr  0.11 csys =  0.62 CPU)
Result: FAIL
```
Since the `gcc` version I was testing with dates back to 2007, I decided to try with the system's default C-compiler, `clang-10`.
```
$ clang --version                                                               >
OpenBSD clang version 10.0.1 
Target: amd64-unknown-openbsd6.9
Thread model: posix
InstalledDir: /usr/bin

$ sh ./Configure -des -Dusedevel -Dcc=clang -Accflags='-DPERL_RC_STACK -DDEBUG_LEAKING_SCALARS'; make test_prep
```
With this more recent `clang`, the failing tests PASSed.
```
$ cd t; ./perl harness -v ../lib/ExtUtils/t/Embed.t; cd -

ok 1
ok 2
ok 3
ok 4
ok 5
ok 6
ok 7
ok 8
ok 9
ok 10 # system returned 0
ok
All tests successful.
Files=1, Tests=10,  1 wallclock secs ( 0.01 usr  0.02 sys +  0.55 cusr  0.36 csys =  0.94 CPU)
Result: PASS
```
It would be good if someone could test these config options on an up-to-date OpenBSD with both `gcc` and `clang`.  We can then try to evaluate the source of the test failures.  cc: @afresh1

Noting the 2 instances of `ld: error: undefined symbol ...`:
```
$ cd t && \
gcc -o embed_test -I..  -std=gnu99 -DNO_LOCALE_NUMERIC -DNO_LOCALE_COLLATE -DNO_LOCALE_MONETARY -DNO_LOCALE_TIME -DNO_LOCALE_MESSAGES -DLIBC_HANDLES_MISMATCHED_CTYPE -DDEBUG_LEAKING_SCALARS -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include   embed_test.c -L.. -lperl   -Wl,-E  -fstack-protector-strong -L/usr/local/lib  -lperl -lpthread -lm -lutil -lc; cd -

ld: error: undefined symbol: Perl_more_sv
>>> referenced by embed_test.c
>>>               /tmp//cc2IxT9I.o:(S_new_SV)

ld: error: undefined symbol: PL_sv_serial
>>> referenced by embed_test.c
>>>               /tmp//cc2IxT9I.o:(S_new_SV)
>>> referenced by embed_test.c
>>>               /tmp//cc2IxT9I.o:(S_new_SV)
collect2: ld returned 1 exit status
```
```
[openbsd69: perl] $ ack -l '(Perl_more_sv)' .              
sv_inline.h
embed.h
proto.h
sv.c
[openbsd69: perl] $ ack -l '(PL_sv_serial)' .              
sv_inline.h
embedvar.h
makedef.pl
lib/Devel/PPPort.pm
...
```
Further analysis needs more C-foo than I have.

I had a look at this and reproduced it locally.  libperl.a looks fine:
```
$ nm libperl.a | grep PL_sv_serial
...
00006738 B PL_sv_serial
...
```
So what do we end up linking against? I edited `Embed.t` to add `-Wl,-t` to the compiler invocation:
```
# eg++ -o embed_test -Wl,-t -I..  -DNO_LOCALE_NUMERIC -DNO_LOCALE_COLLATE -DNO_LOCALE_MONETARY -DNO_LOCALE_TIME -DNO_LOCALE_MESSAGES -DLIBC_HANDLES_MISMATCHED_CTYPE -DPERL_RC_STACK -DDEBUG_LEAKING_SCALARS -fwrapv -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include   embed_test.c -L.. -lperl   -Wl,-E  -fstack-protector-strong -L/usr/local/lib  -lperl -lpthread -lm -lutil -lc
ld: error: undefined symbol: PL_sv_serial
>>> referenced by embed_test.c
>>>               /tmp//ccIahN7A.o:(S_new_SV)
>>> referenced by embed_test.c
>>>               /tmp//ccIahN7A.o:(S_new_SV)
collect2: error: ld returned 1 exit status
# /usr/lib/crt0.o

# /usr/lib/crtbegin.o

# /tmp//ccIahN7A.o

# /usr/lib/libperl.so.23.0

# /usr/lib/libperl.so.23.0

# /usr/lib/libpthread.so.27.0

# /usr/lib/libutil.so.16.0

# /usr/local/lib/libestdc++.so.20.0

# /usr/lib/libm.so.10.1

# /usr/lib/libc.so.97.0

# /usr/lib/libc.so.97.0

# /usr/lib/crtend.o

not ok 1
# embed_test = ./embed_test
not ok 10 # system returned -1
Failed 10/10 subtests 
```
So we're linking against the system libperl, which of course wasn't built with `-DDEBUG_LEAKING_SCALARS` and doesn't define `PL_sv_serial`.

So while 75acd14e43f2ffb698fc7032498f31095b56adb5 revealed this problem, it didn't cause the problem.

> I need to look into it further, perl itself seems to link fine.

The perl executable links the library via direct reference:
```
eg++ -o perl -Wl,-E  -fstack-protector-strong -L/usr/local/lib  perlmain.o   libperl.a `cat ext.libs` -lpthread -lm -lutil -lc
```
So I see two issues directly related to the test here:
- the link process for embed_test doesn't match that of the perl executable, so we might link the wrong library in (and do in this case)
- the test itself doesn't detect that we linked the wrong library, it's fairly simple and doesn't do anything that involves the layout of the perl structure

I can see a couple of issues not directly related to the failures:
- `S_new_SV` should probably be PERL_STATIC_INLINE, or maybe better just defined in `sv.c` and exported, ie. E in embed.fnc, to allow simpler breakpoints
- the `S_new_SV` file, line, func arguments should be `PERL_ARG_UNUSED()`:
```
In file included from ../../perl.h:7870,
                 from B.xs:13:
../../sv_inline.h: In function 'SV* S_new_SV(const char*, int, const char*)':
../../sv_inline.h:75:28: warning: unused parameter 'file' [-Wunused-parameter]
   75 | S_new_SV(pTHX_ const char *file, int line, const char *func)
      |                ~~~~~~~~~~~~^~~~
../../sv_inline.h:75:38: warning: unused parameter 'line' [-Wunused-parameter]
   75 | S_new_SV(pTHX_ const char *file, int line, const char *func)
      |                                  ~~~~^~~~
../../sv_inline.h:75:56: warning: unused parameter 'func' [-Wunused-parameter]
   75 | S_new_SV(pTHX_ const char *file, int line, const char *func)
```

As https://perl5.test-smoke.org has come back to life (thanks to Leo, Olaf, *et al.*), I've been able to gather more data about this problem -- and am glad that the problem was **partially** solved by @richardleach's  https://github.com/Perl/perl5/commit/c79fe2b42ae2a540552f87251aa0e36a060dd584 back in June.  I noticed that after a string of 5 smoke-test run failures on my OpenBSD-6.9 VM between April 6 and June 11, no failures have been observed since.  I bisected with the following invocation:
```
perl Porting/bisect.pl -Dcc=gcc \
    -Accflags='-DPERL_RC_STACK -DDEBUG_LEAKING_SCALARS' \
--start a902d92a78 \
--end 01f2355587 \
--expect-fail \
--target lib/ExtUtils/t/Embed.t
```
... and found that the first "bad" (really, good, in this case) commit was :+1: 
```
c79fe2b42ae2a540552f87251aa0e36a060dd584 is the first bad commit
commit c79fe2b42ae2a540552f87251aa0e36a060dd584
Author: Richard Leach <richardleach@users.noreply.github.com>
Date:   Sat May 11 13:26:27 2024 +0000
Commit:     Richard Leach <rich+perl@hyphen-dash-hyphen.info>
CommitDate: Wed Jun 12 12:44:21 2024 +0100

    S_new_SV: args unused, static inline & defined, rename with Perl_ prefix
```
This build used `gcc` but was not a threaded build.  Today I decided to confirm that I got a PASS on a threaded build.  Today I configured with:
```
$ sh ./Configure -des -Dusedevel -Duseithreads -Dcc=gcc
```
... then built and tested.
```
$ cd t; ./perl harness -v ../lib/ExtUtils/t/Embed.t; cd -                 
ld: error: undefined symbol: Perl_croak_nocontext
>>> referenced by embed_test.c
>>>               /tmp//ccTDHCnn.o:(Perl_croak_memory_wrap)
collect2: ld returned 1 exit status

not ok 1
not ok 10 # system returned -1
Failed 10/10 subtests 

Test Summary Report
-------------------
../lib/ExtUtils/t/Embed.t (Wstat: 0 Tests: 2 Failed: 2)
  Failed tests:  1, 10
  Parse errors: Tests out of sequence.  Found (10) but expected (2)
                Bad plan.  You planned 10 tests but ran 2.
Files=1, Tests=2,  1 wallclock secs ( 0.00 usr  0.05 sys +  0.45 cusr  0.18 csys =  0.68 CPU)
Result: FAIL
Finished test run at Sun Sep  1 12:11:21 2024.
```
Note that the very first entry in this issue back in April was reporting an *unthreaded* build.

I'm hoping that we can get some smoke-tests on more up-to-date OpenBSDs (7.5) with more modern `gcc`s, unthreaded and threaded, with and without `-Accflags='-DPERL_RC_STACK -DDEBUG_LEAKING_SCALARS'` so that we can get a better handle on this problem.

--- a/embed.fnc
+++ b/embed.fnc
@@ -800,6 +800,10 @@ Adp	|SV *	|amagic_deref_call					\
 p	|bool	|amagic_is_enabled					\
 				|int method
 
+CTdp	|void	|api_version_assert					\
+				|size_t interp_size			\
+				|NULLOK void *v_my_perl 		\
+				|NN const char *api_version
 ETXip	|void	|append_utf8_from_native_byte				\
 				|const U8 byte				\
 				|NN U8 **dest

--- a/embed.h
+++ b/embed.h
@@ -61,6 +61,7 @@
 # define SvUV_nomg(a)                           Perl_SvUV_nomg(aTHX_ a)
 # define amagic_call(a,b,c,d)                   Perl_amagic_call(aTHX_ a,b,c,d)
 # define amagic_deref_call(a,b)                 Perl_amagic_deref_call(aTHX_ a,b)
+# define api_version_assert                     Perl_api_version_assert
 # define apply_attrs_string(a,b,c,d)            Perl_apply_attrs_string(aTHX_ a,b,c,d)
 # define apply_builtin_cv_attributes(a,b)       Perl_apply_builtin_cv_attributes(aTHX_ a,b)
 # define atfork_child                           Perl_atfork_child

--- a/lib/ExtUtils/t/Embed.t
+++ b/lib/ExtUtils/t/Embed.t
@@ -103,13 +103,26 @@ if ($^O eq 'VMS') {
    elsif ($^O eq 'os390' && $Config{usedl}) {
     push(@cmd,"-L$lib", ldopts());
    } else { # Not MSWin32 or OS/390 (z/OS) dynamic.
-    push(@cmd,"-L$lib",'-lperl');
+    my $ldopts = ldopts();
+    if ($^O eq 'openbsd' && $Config{useshrplib} eq "false") {
+        # see github #22125
+        # with OpenBSD, the packaged gcc (tries to) link
+        # against the system libperl, this will be fine once
+        # this perl gets installed, but that's not so good when
+        # testing against the uninstalled perl.
+        # This also matches how Makefile.SH links the perl executable
+        push @cmd, "$lib/libperl.a";
+        $ldopts =~ s/ -lperl\b//;
+    }
+    else {
+        push(@cmd,"-L$lib",'-lperl');
+    }
     local $SIG{__WARN__} = sub {
 	warn $_[0] unless $_[0] =~ /No library found for .*perl/
     };
     push(@cmd, '-Zlinker', '/PM:VIO')	# Otherwise puts a warning to STDOUT!
 	if $^O eq 'os2' and $Config{ldflags} =~ /(?<!\S)-Zomf\b/;
-    push(@cmd,ldopts());
+    push(@cmd, $ldopts);
    }
 
    if ($^O eq 'aix') { # AIX needs an explicit symbol export list.
@@ -196,7 +209,7 @@ int main(int argc, char **argv, char **env) {
     perl_construct(my_perl);
     PL_exit_flags |= PERL_EXIT_WARN;
 
-    PERL_API_VERSION_CHECK;
+    PERL_API_VERSION_ASSERT;
 
     my_puts("ok 3");
 

--- a/lib/ExtUtils/t/Embed.t
+++ b/lib/ExtUtils/t/Embed.t
@@ -196,6 +196,8 @@ int main(int argc, char **argv, char **env) {
     perl_construct(my_perl);
     PL_exit_flags |= PERL_EXIT_WARN;
 
+    PERL_API_VERSION_CHECK;
+
     my_puts("ok 3");
 
     perl_parse(my_perl, NULL, (sizeof(cmds)/sizeof(char *))-1, (char **)cmds, env);

--- a/perl.h
+++ b/perl.h
@@ -9276,6 +9276,14 @@ END_EXTERN_C
 #  define PERL_STACK_REALIGN
 #endif
 
+#ifdef MULTIPLICITY
+#  define PERL_API_VERSION_ASSERT \
+  Perl_api_version_assert(sizeof(PerlInterpreter), aTHX, PERL_API_VERSION_STRING)
+#else
+#  define PERL_API_VERSION_ASSERT \
+  Perl_api_version_assert(sizeof(PerlInterpreter), NULL, PERL_API_VERSION_STRING)
+#endif
+
 /*
 
    (KEEP THIS LAST IN perl.h!)

--- a/pod/perldelta.pod
+++ b/pod/perldelta.pod
@@ -309,6 +309,11 @@ made:
 
 XXX
 
+=item *
+
+When testing embedding, add a sanity check to ensure the C<libperl> we
+link against matches the perl we are building.  [GH #22125]
+
 =back
 
 =head1 Platform Support
@@ -357,6 +362,11 @@ L</Modules and Pragmata> section.
 =item XXX-some-platform
 
 XXX
+
+=item OpenBSD
+
+When testing embedding, ensure we link against the correct static
+libperl. [GH #22125]
 
 =back
 

--- a/pod/perldiag.pod
+++ b/pod/perldiag.pod
@@ -4071,6 +4071,11 @@ See L</500 Server error>.
 by a missing delimiter on a string or pattern, because it eventually
 ended earlier on the current line.
 
+=item Mismatch between expected and libperl %s
+
+(F) For an embedded perl, the perl headers and configuration you built
+your binary against don't match the library you've linked with.
+
 =item Mismatched brackets in template
 
 (F) A pack template could not be parsed because pairs of C<[...]> or

--- a/proto.h
+++ b/proto.h
@@ -123,6 +123,11 @@ Perl_amagic_is_enabled(pTHX_ int method)
         __attribute__visibility__("hidden");
 #define PERL_ARGS_ASSERT_AMAGIC_IS_ENABLED
 
+PERL_CALLCONV void
+Perl_api_version_assert(size_t interp_size, void *v_my_perl, const char *api_version);
+#define PERL_ARGS_ASSERT_API_VERSION_ASSERT     \
+        assert(api_version)
+
 PERL_CALLCONV SSize_t
 Perl_apply(pTHX_ I32 type, SV **mark, SV **sp)
         __attribute__visibility__("hidden");

--- a/util.c
+++ b/util.c
@@ -5656,6 +5656,52 @@ S_xs_version_bootcheck(pTHX_ SSize_t items, SSize_t ax, const char *xs_p,
     }
 }
 
+/*
+=for apidoc api_version_assert
+
+Used by the PERL_API_VERSION_CHECK macro to compare the perl the
+object was built with and the perl that C<libperl> was built with.
+
+This can be used to ensure that these match and produces a more
+diagnosable than random crashes and mis-behaviour.
+
+=cut
+*/
+
+void
+Perl_api_version_assert(size_t interp_size, void *v_my_perl,
+                        const char *api_version) {
+    dTHX;
+
+    PERL_ARGS_ASSERT_API_VERSION_ASSERT;
+
+    if (interp_size != sizeof(PerlInterpreter)) {
+        /* detects various types of configuration mismatches */
+        /* diag_listed_as: Mismatch between expected and libperl %s */
+        Perl_croak(aTHX_
+                   "Mismatch between expected and libperl interpreter structure size %zd vs %zd",
+                   interp_size, sizeof(PerlInterpreter));
+    }
+    if (
+#ifdef MULTIPLICITY
+      v_my_perl != my_perl
+#else
+      v_my_perl != NULL
+#endif
+        ) {
+        /* detect threads vs non-threads mismatch */
+        /* diag_listed_as: Mismatch between expected and libperl %s */
+        Perl_croak(aTHX_
+                   "Mismatch between expected and libperl interpreter pointer");
+    }
+    if (strNE(api_version, PERL_API_VERSION_STRING)) {
+        /* diag_listed_as: Mismatch between expected and libperl %s */
+        Perl_croak(aTHX_
+                   "Mismatch between expected and libperl API versions %s vs %s",
+                   api_version, PERL_API_VERSION_STRING);
+    }
+}
+
 PERL_STATIC_INLINE bool
 S_gv_has_usable_name(pTHX_ GV *gv)
 {

